internal/launcher/launcher.go

```diff
@@ -0,0 +1,158 @@
+)
+
+// Execute runs the k6 command.
+func Execute() {
+	gs := state.NewGlobalState(context.Background())
+
+	newLauncher(gs).launch()
+}
+
+// launcher is a k6 launcher
+type launcher struct {
+	gs *state.GlobalState
+	// function to fall back if binary provisioning is not required
+	fallback func(gs *state.GlobalState)
+	// function to provision a k6 binary that satisfies the dependencies
+	provision func(*state.GlobalState, k6deps.Dependencies) (string, string, error)
+	// function to execute k6 binary
+	run func(*state.GlobalState, string) (int, error)
+}
+
+func newLauncher(gs *state.GlobalState) *launcher {
+	return &launcher{
+		gs:        gs,
+		fallback:  k6Cmd.ExecuteWithGlobalState,
+		provision: k6buildProvision,
+		run:       runK6Cmd,
+	}
+}
+
+// launch executes k6 either by launching a provisioned binary or defaulting to the
+// current binary it this is not necessary.
+// If the fhe fallback is called, it can exit the process so don't assume it will return
+func (l *launcher) launch() {
+	// if binary provisioning not enabled, continue with regular k6 execution path
+	if !l.gs.Flags.BinaryProvisioning {
+		l.gs.Logger.Debug("binary provisioning disabled")
+		l.fallback(l.gs)
+		return
+	}
+
+	l.gs.Logger.Info("trying to provision binary")
+
+	deps, err := k6deps.Analyze(newDepsOptions(l.gs, l.gs.CmdArgs[1:]))
+	if err != nil {
+		l.gs.Logger.
+			WithError(err).
+			Error("failed to analyze dependencies, can't try binary provisioning, please report this issue")
+		l.gs.OSExit(1)
+	}
+
+	// binary provisioning enabled but not required by this command
+	// continue with regular k6 execution path
+	if !isCustomBuildRequired(build.Version, deps) {
+		l.gs.Logger.
+			Debug("binary provisioning not required")
+		l.fallback(l.gs)
+		return
+	}
+
+	l.gs.Logger.
+		WithField("deps", deps).
+		Info("dependencies identified, binary provisioning required")
+
+	l.launchCustomBuild(deps)
+}
+
+func (l *launcher) launchCustomBuild(deps k6deps.Dependencies) {
+	// get the k6 binary from the build service
+	binPath, versions, err := l.provision(l.gs, deps)
+	if err != nil {
+		l.gs.Logger.
+			WithError(err).
+			Error("failed to fetch a binary with required dependencies, please report this issue")
+		l.gs.OSExit(1)
+		// in tests calling l.gs.OSExit does not ends execution so we have to return
+		return
+	}
+
+	l.gs.Logger.
+		Info("k6 has been provisioned with version(s) ", versions)
+
+	l.gs.Logger.Debug("launching provisioned k6 binary")
+
+	// execute provisioned binary
+	if rc, err := l.run(l.gs, binPath); err != nil {
+		l.gs.Logger.Error(err)
+		l.gs.OSExit(rc)
+	}
+}
+
+// runK6Cmd runs the k6 binary passing the original arguments
+func runK6Cmd(gs *state.GlobalState, binPath string) (int, error) {
+	cmd := exec.CommandContext(gs.Ctx, binPath, gs.CmdArgs[1:]...) //nolint:gosec
+	cmd.Stderr = gs.Stderr
+	cmd.Stdout = gs.Stdout
+	cmd.Stdin = gs.Stdin
+
+	// disable binary provisioning to avoid a provisioning loop
```

Comment:
Why do we need this? It looks like a "patch", other than how it should really behave. Is it because so far we're not able to correctly detect if the dependencies (extensions used) are really support by the current binary? 🤔 

Any other reason?

Reply:
The binary we are providing is a `k6` binary that is also capable of binary provisioning. If we don't disable the option, it will execute the same path as the original one: scan the input script, detect the dependencies, call binary provisioning, execute the binary ... which will repeat the loop.